### ISSUE TITLE
Url not updated after deleting last empty thought

### ISSUE THREAD

**Issue by raineorshine** [cybersemics/em#912]
When the last empty thought in the root is deleted, the cursor changes but the url is not updated.

## Steps to Reproduce

Create an empty thought and delete it.

## Current Behavior

Url is `http://localhost:3000/~/` (i.e. empty thought in the root, which we would only expect if the thought still exited)

## Expected Behavior

Url should be updated to `http://localhost:3000`

**Comment by shresthabijay** [comment]
@raineorshine This issue is very tricky.  In `updateUrlHistory` `decoded` path returned by `decodeThoughtsUrl` for location `http://localhost:3000/~/` is `null` after thought is deleted. 

https://github.com/cybersemics/em/blob/0e95af78232b43f5b52af8e054616cf5e5e753c9/src/selectors/decodeThoughtsUrl.ts#L25-L35

After new thought is deleted  `pathUnranked` for `http://localhost:3000/~/` is `['']`. Here you can see in line `25` that `thoughtsLoaded` will be `false` because `getThought(state, '')`  returns `null` as only empty thought in the tree is already deleted. So `decodeThoughtsUrl`  returns `null`  for `http://localhost:3000/~/` right after new thought is deleted.

Now let's have a look at `updateUrlHistory`.

https://github.com/cybersemics/em/blob/0e95af78232b43f5b52af8e054616cf5e5e753c9/src/redux-middleware/updateUrlHistory.ts#L31

As you can see  `contextDecoded` becomes `[ROOT_TOKEN]` since `decoded.path` is null. 

https://github.com/cybersemics/em/blob/0e95af78232b43f5b52af8e054616cf5e5e753c9/src/redux-middleware/updateUrlHistory.ts#L27

Also `context` is `[ROOT_TOKEN]` when path (in this case cursor) is `null`. 

https://github.com/cybersemics/em/blob/0e95af78232b43f5b52af8e054616cf5e5e753c9/src/redux-middleware/updateUrlHistory.ts#L34

Since both `context` and `contextDecoded` are same, the further execution is stopped. Thus url doesn't get updated.

We cannot simply set `contextDecoded` to `[ROOT_TOKEN]`  if `decoded.path` is `null`. Instead we should allow further execution and update url with new incoming context.

```js
if (decoded.path && equalArrays(pathToContext(decoded.path), context) && decoded.contextViews[encoded] === (contextViews || state.contextViews)[encoded]) return
```

I tried this but it I ran into some issues. Turns out `decodeThoughtsUrl` is always returning `null` for any `pathname` on first few execution after app is started. It seems like `updateUrlHistory` is being called before thoughts are loaded into the `state`. Since after iterative loading PR things have changed on how thoughts are loaded, I will need your help on this.

**Comment by shresthabijay** [comment]
https://github.com/cybersemics/em/blob/0e95af78232b43f5b52af8e054616cf5e5e753c9/src/reducers/updateThoughts.ts#L137

On further inspection this logic is being used to populate correct cursor from the url on first load. But it always returns `null`  due to following logic in the `decodeThoughtsUrl`.

https://github.com/cybersemics/em/blob/0e95af78232b43f5b52af8e054616cf5e5e753c9/src/selectors/decodeThoughtsUrl.ts#L25-L35

Also, this logic will prevent url to update when only thought (non empty) in the root context is deleted.

- Create a new thought `a` in the root context.
- Delete thought `a`.

Since new `cursor` is will be `null` this logic that is intended to run on first load gets executed. Here `decodeThoughtsUrl` returns a valid `decoded.path`.

https://github.com/cybersemics/em/blob/0e95af78232b43f5b52af8e054616cf5e5e753c9/src/selectors/decodeThoughtsUrl.ts#L25-L35

Here `thoughtsLoaded` will be `true` even though the thought `a`  has been deleted.  That is because `a` still exits on context `['=archived']` . Also `rankThoughtsFirstMatch` seems to return a valid `path` for any given non existent `context`. 

This will reset the `cursor` back to decoded `path` from the url. Thus causing both `cursor` and url to get properly updated.

**Comment by raineorshine** [comment]
> After new thought is deleted `pathUnranked` for `http://localhost:3000/~/` is `['']`. Here you can see in line `25` that `thoughtsLoaded` will be `false` because `getThought(state, '')` returns `null` as only empty thought in the tree is already deleted. So `decodeThoughtsUrl` returns `null` for `http://localhost:3000/~/` right after new thought is deleted.

I see. The logic added to prevent loading an invalid cursor (the [`getThought`](https://github.com/cybersemics/em/blob/0e95af78232b43f5b52af8e054616cf5e5e753c9/src/selectors/decodeThoughtsUrl.ts#L25) check) causes `decodeThoughtsUrl` to return `null` instead of the actual path that it decoded.

> Since both `context` and `contextDecoded` are same, the further execution is stopped. Thus url doesn't get updated.

Exactly.

> We cannot simply set `contextDecoded` to `[ROOT_TOKEN]` if `decoded.path` is `null`. Instead we should allow further execution and update url with new incoming context.
> 
> ```js
> if (decoded.path && equalArrays(pathToContext(decoded.path), context) && decoded.contextViews[encoded] === (contextViews || state.contextViews)[encoded]) return
> ```
> 
> I tried this but it I ran into some issues. Turns out `decodeThoughtsUrl` is always returning `null` for any `pathname` on first few execution after app is started. It seems like `updateUrlHistory` is being called before thoughts are loaded into the `state`. Since after iterative loading PR things have changed on how thoughts are loaded, I will need your help on this.

That's right, on first load of a non-root url, the thoughts won't have been populated yet. `updateUrlHistory` needs to fail gracefully in this case. Requiring a non-null `decoded.path` in this conditional breaks in cases where the decoded path is correctly `null`.

I think what is needed is an `exists` option to optionally require that `decodeThoughtsUrl` returns `null` if the `Path` does not exist in the thought cache. I'll try it out.

**Comment by raineorshine** [comment]
Yeah, not checking if the path exists in `updateUrlHistory` fixes it: #916

**Comment by raineorshine** [comment]
> Also, this logic will prevent url to update when only thought (non empty) in the root context is deleted.

#916 fixes this, too. Note that this only occurs if `a` is deleted quickly, before the Editable's `thoughtChangeHandler` is flushed. If you wait a second after creating `a` and then delete, you get a cursor bug (#911).
